src/DataSource.ts

```diff
@@ -68,11 +69,23 @@ export class DataSource extends DataSourceApi<GrafanaQuery, GenericOptions> {
     });
   }
 
-  metricFindQuery(query: string, options?: any, type?: string): Promise<MetricFindValue[]> {
-    const interpolated = {
-      type,
-      target: getTemplateSrv().replace(query, undefined, 'regex'),
-    };
+  metricFindQuery(legacyOrNew: VariableQuery | string, options?: any, type?: string): Promise<MetricFindValue[]> {
+    let query: VariableQuery;
+    if (typeof legacyOrNew === 'object') {
+      query = legacyOrNew;
+    } else {
+      query = { query: legacyOrNew, asJson: false };
```

Comment:
Ok first of all, these are all great points you're making and I'm totally happy to change approaches. I've put a lot of thought on how to deal with this so let me explain and you can tell me what you think we can do about it...

The purpose of this section is specifically to reconcile the new style (an actual object where we can put keys/values for this feature and anything else we would want in the future) with the existing dashboard schemas that everyone currently has stored in a regular string with this query field.

The existing schema:
```
...
`"templating": {
    "list": [
      {
        "allValue": null,
        "current": {},
        "datasource": null,
        "definition": "{\"Thisisatest\":\"test\"}",
        "description": null,
        "error": null,
        "hide": 0,
        "includeAll": false,
        "label": null,
        "multi": false,
        "name": "query0",
        "options": [],
        "query": "{\"Thisisatest\":\"test\"}",
        "refresh": 0,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }
    ]
  }...
```

vs the new (which I have now changed `asJson` to `format`):
```
...
"templating": {
    "list": [
      {
        "allValue": null,
        "current": {},
        "datasource": null,
        "definition": "{\"Thisisatest\":\"test\"} (JSON)",
        "description": null,
        "error": null,
        "hide": 0,
        "includeAll": false,
        "label": null,
        "multi": false,
        "name": "query0",
        "options": [],
        "query": {
          "format": "json",
          "query": "{\"Thisisatest\":\"test\"}",
        },
        "refresh": 0,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }...
```

If we don't have some way to convert the old "string" style to an actual object, everyone will have to re-create their template variable queries. By checking if we've received a plain string, we can convert it to the new. If they make any change to their template query and then resave their dashboard, it will be saved in the new format.

The downside is this annoying check, and it'll probably be years before we can assume everyone has touched their template variable queries at least once to remove the check. I don't know what else to do other than to piss everyone off by nulling out their existing dashboard template variable query strings when they update.

On the plus side, in the future it'll be way easier to expand this query object, because it's non-breaking to add another key (for example, maybe someday we will we want to add support to include the date range).

I did think about ways we could keep the traditional old string style and still add this new support, like maybe prepend the string with some sort of extra data to signify the type, but I think this really limits us and doesn't feel nicer than that check. Ie. we could make something that we can parse out of the string:

```
        "query": "ASJSON: {\"Thisisatest\":\"test\"}",
```

But then that still requires some sort of check to see if it's old or new, and it's back to the same problem as before where we're parsing the string to see if it's valid JSON or not at query time. So I figured we might as well just switch to an actual type (VariableQuery) and suck up the check to help people upgrade. In a future release maybe we can tell everyone that they must have updated their template variable queries since this version and re-saved their dashboards at least once, to remove that check.

Reply:
Ah right. Can we keep it simple, so the code would stay as is so people can migrate their dashboards.

And will drop the condition in the next "minor" version (because we're still at `0.x.y`).

Reply:
Yes I think that's good, or maybe we could drop it in the first release of this plugin after Grafana 8.0 comes out this year. I find at those times people have a high tolerance for potentially breaking changes if you don't follow proper upgrade path (ie. most people who use the UI we can just tell them that they must flick the new JSON switch back and forth once and re-save their dashboards).

Reply:
Can you please prepare PR after this gets merged with the drop?